### PR TITLE
Serialise base members and nlohmann::json with msgpack-c

### PR DESCRIPTION
**Pull request by eddyashton**
Some of our uses of `MSGPACK_DEFINE` were not serialising members inherited from the base type. In particular we were not fully serialising `OpenProposal`s, resulting in differences in the KV across replicas. This resulted in #214.

The fix is simple - adding `MSGPACK_BASE(T)` entries. In the case of `OpenProposal` I decided it was simpler to remove the inheritance (the fields it shares with `Proposal::In` are coincidental, not a logical hierarchy). To serialise the `nlohmann::json` field I needed to define msgpack-c's `pack` and `convert` structs, and using _that_ we can replace the special case handling of json objects in `msgpackserialise.h`.

I've added `msgpack_serialization_test` which roundtrips examples of all these types, but we don't have a way to enforce that new fields are correctly serialised.

### CODE REVIEW

File: src/node/rpc/test/membervoting_test.cpp [hunk 1]
```diff
@@ -315,49 +338,77 @@ TEST_CASE("Add new members until there are 7, then reject")
...
+    CHECK(initial_read.result.parameter == new_member.cert);
+
+    // vote as second member
+    Script vote_ballot(fmt::format(
+      R"xxx(
         local tables, calls = ...
         local n = 0
         tables["members"]:foreach( function(k, v) n = n + 1 end )
-        if n < 8 then
+        if n < {} then
```
Comment (achamayou) [inline src/node/rpc/test/membervoting_test.cpp]: ?

Comment (eddyashton) [reply src/node/rpc/test/membervoting_test.cpp]: libfmt substitutes in `max_members`, so this still becomes `if n < 8` before Lua sees it. But if we change `max_members`, the script still works.

Comment (achamayou) [reply src/node/rpc/test/membervoting_test.cpp]: Oh I see!

**Pull request merged**